### PR TITLE
run type core components

### PR DESCRIPTION
**Pull request by axenteoctavian**
## Reasoning behind the pull request
- Needed run type for core components (normal or sovereign)
- 
- 
  
## Proposed changes
- implement `RunTypeCoreComponents`
- 
- 

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: factory/runType/sovereignRunTypeCoreComponents.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"github.com/multiversx/mx-chain-core-go/core/check"
+
+	"github.com/multiversx/mx-chain-go/errors"
+	"github.com/multiversx/mx-chain-go/process/rating"
+	"github.com/multiversx/mx-chain-go/sharding"
+)
+
+type sovereignRunTypeCoreComponentsFactory struct {
```
Comment (mariusmihaic) [inline factory/runType/sovereignRunTypeCoreComponents.go:11]: We need to have `IsInterfaceNil` as method mandatory. Inheritance is not the way to go with that

File: node/chainSimulator/components/coreComponents_test.go [hunk 2]
```diff
@@ -140,13 +140,22 @@ func createArgsCoreComponentsHolder() ArgsCoreComponentsHolder {
...
-			return sharding.NewNodesSetup(nodesFilePath, addressPubkeyConverter, validatorPubkeyConverter, genesisMaxNumShards)
-		},
-		CreateRatingsData: func(arg rating.RatingsDataArg) (process.RatingsInfoHandler, error) {
-			return rating.NewRatingsData(arg)
-		},
+		RunTypeCoreComponents:       runTypeCoreComponents,
+	}
+}
+
+func createRunTypeCoreComponents() (mainFactory.RunTypeCoreComponentsHolder, error) {
```
Comment (mariusmihaic) [inline node/chainSimulator/components/coreComponents_test.go:147]: This is copy-pasted with return error. In test files, we could remove the error, perhaps only have `t *tesing.T` as input param

File: testscommon/components/components.go [hunk 3]
```diff
@@ -82,8 +81,42 @@ const DummySk = "cea01c0bf060187d90394802ff223078e47527dc8aa33a922744fb1d06029c4
...
+	return managedRunTypeCoreComponents
+}
+
+// GetSovereignRunTypeCoreComponents -
+func GetSovereignRunTypeCoreComponents() factory.RunTypeCoreComponentsHolder {
+	runTypeCoreComponentsFactory := runType.NewRunTypeCoreComponentsFactory()
+	sovRunTypeCoreComponentsFactory, _ := runType.NewSovereignRunTypeCoreComponentsFactory(runTypeCoreComponentsFactory)
+	managedRunTypeCoreComponents, err := runType.NewManagedRunTypeCoreComponents(sovRunTypeCoreComponentsFactory)
+	if err != nil {
+		log.Error("getSovereignRunTypeCoreComponents NewManagedRunTypeCoreComponents", "error", err.Error())
```
Comment (mariusmihaic) [inline testscommon/components/components.go:106]: Copy pasted log error here and below ?

File: factory/runType/runTypeCoreComponents_test.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+func TestRunTypeCoreComponentsFactory_Close(t *testing.T) {
+	t.Parallel()
+
+	rccf := runType.NewRunTypeCoreComponentsFactory()
+	require.NotNil(t, rccf)
+
+	rcc := rccf.Create()
+	require.NotNil(t, rcc)
```
[the remark this answers is not in the document]
Comment (axenteoctavian) [reply factory/runType/runTypeCoreComponents_test.go:37]: I need to create the components to be able to close

File: factory/runType/runTypeCoreComponentsHandler.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+var _ factory.RunTypeCoreComponentsHolder = (*managedRunTypeCoreComponents)(nil)
+
+type managedRunTypeCoreComponents struct {
+	*runTypeCoreComponents
+	factory                  runTypeCoreComponentsCreator
+	mutRunTypeCoreComponents sync.RWMutex
+}
+
+// NewManagedRunTypeCoreComponents returns a news instance of managedRunTypeCoreComponents
+func NewManagedRunTypeCoreComponents(rcf runTypeCoreComponentsCreator) (*managedRunTypeCoreComponents, error) {
```
[the remark this answers is not in the document]
Comment (axenteoctavian) [reply factory/runType/runTypeCoreComponentsHandler.go:25]: I cannot rename the argument, `runTypeCoreComponentsFactory` is a struct.

File: factory/runType/runTypeCoreComponentsFactory.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+type runTypeCoreComponentsFactory struct {
+}
+
+// NewRunTypeCoreComponentsFactory will return a new instance of runTypeCoreComponentsFactory
+func NewRunTypeCoreComponentsFactory() *runTypeCoreComponentsFactory {
+	return &runTypeCoreComponentsFactory{}
+}
+
+func (rccf *runTypeCoreComponentsFactory) Create() *runTypeCoreComponents {
```
Comment (mariusmihaic) [inline factory/runType/runTypeCoreComponentsFactory.go:16]: Missing comm

File: node/chainSimulator/components/testOnlyProcessingNode_test.go [hunk 7]
```diff
@@ -165,6 +183,13 @@ func TestNewTestOnlyProcessingNode(t *testing.T) {
 		require.Error(t, err)
 		require.Nil(t, node)
 	})
+	t.Run("CreateRunTypeComponents failure should error", func(t *testing.T) {
```
Comment (mariusmihaic) [inline node/chainSimulator/components/testOnlyProcessingNode_test.go:186]: CreateRunTypeComponents fails or  CreateRunType**CORE**Components ?

File: factory/runType/runTypeCoreComponentsHandler.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply factory/runType/runTypeCoreComponentsHandler.go:25]: I meant the interface renaming from `runTypeCoreComponentsCreator` -> `runTypeCoreComponentsFactory`.

W/e, can be left as it is

[the remark this answers is not in the document]
Comment (axenteoctavian) [reply factory/runType/runTypeCoreComponentsHandler.go:25]: Yes, but `runTypeCoreComponentsFactory` is already a struct defined, and I cannot have an interface with the same name

**Review by mariusmihaic — APPROVED** [review]

File: factory/runType/runTypeCoreComponentsHandler.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply factory/runType/runTypeCoreComponentsHandler.go:25]: Yes, that's why interfaces usually have `handler` at the end of the name.
We can leave it as it is

**Review by sasurobert — APPROVED** [review]

**Pull request merged**